pkg/controller/topology_test.go

```diff
@@ -146,6 +146,210 @@ func TestGenerateVolumeNodeAffinity(t *testing.T) {
 	}
 }
 
+func TestStatefulSetSpreading(t *testing.T) {
+	nodeLabels := []map[string]string{
+		{"com.example.csi/zone": "zone1", "com.example.csi/rack": "rackA"},
+		{"com.example.csi/zone": "zone2", "com.example.csi/rack": "rackB"},
+		{"com.example.csi/zone": "zone3", "com.example.csi/rack": "rackC"},
+		{"com.example.csi/zone": "zone4", "com.example.csi/rack": "rackD"},
+	}
+	var topologyKeys []map[string][]string
+	keys := map[string][]string{testDriverName: {"com.example.csi/zone", "com.example.csi/rack"}}
+	for i := 0; i < len(nodeLabels); i++ {
+		topologyKeys = append(topologyKeys, keys)
+	}
+
+	testcases := map[string]struct {
+		pvcName           string
+		expectedPreferred []*csi.Topology
+	}{
+		// ordering of segments in preferred array is sensitive to statefulset name: testset
+		// if statefuleset name is changed, make sure expectedPreferred is kept in sync
+		// pvc prefix in pvcName does not have any effect on segment ordering
+		"pvc 1a": {
```

Comment:
Also a test case when allowedTopologies is specified?.  Do we also need to to convert the old zone/zones to AllowedTopologies for migration?

Reply:
Good point - I can add an additional sanity test with `allowedTopologies` specified for completeness. `TestAllowedTopologies` already covers several scenarios for trimming down the `requisiteTerms` based on `allowedTopologies` but it will be good to have a sanity case here as well.

Added point `4b` in https://docs.google.com/document/d/15Q4wXVqbalrVgf_jlGis4QFXN1RY1GMiY-Ek-hpCr-o/edit# for the migration work around topology.

Reply:
Extra test cases with `allowedTopologies` added in latest commit.